field/src/extension/binomial_extension.rs

```diff
@@ -130,41 +130,87 @@ impl<F: BinomiallyExtendable<D>, const D: usize> AbstractField for BinomialExten
 
     #[inline(always)]
     fn square(&self) -> Self {
-        // Specialising mul reduces the computation of c1 from 2 muls
-        // and one add to one mul and a shift
-        *self * *self
+        match D {
+            2 => {
+                let a = self.value;
+
+                Self::from_base_slice(&[a[0].square() + F::W * a[1].square(), a[0] * a[1].double()])
+            }
+            3 => {
+                let a = self.value;
+                let w = F::W;
+
+                let w_a2 = w * a[2];
+
+                let c0 = a[0].square() + (a[1] * w_a2).double();
+                let c1 = w_a2 * a[2] + (a[0] * a[1]).double();
+                let c2 = a[1].square() + (a[0] * a[2]).double();
+
+                Self::from_base_slice(&[c0, c1, c2])
+            }
+            _ => *self * *self,
+        }
     }
 }
 
 impl<F: BinomiallyExtendable<D>, const D: usize> Field for BinomialExtensionField<F, D> {
     type Packing = Self;
-    // Algorithm 11.3.4 in Handbook of Elliptic and Hyperelliptic Curve Cryptography.
     fn try_inverse(&self) -> Option<Self> {
         if self.is_zero() {
             return None;
         }
 
-        // Writing 'a' for self, we need to compute a^(r-1):
-        // r = n^D-1/n-1 = n^(D-1)+n^(D-2)+...+n
-        let mut f = Self::ONE;
-        for _ in 1..D {
-            f = (f * *self).frobenius();
-        }
+        let a = self.value;
+        let w = F::W;
+        match D {
```

Comment:
I'd consider splitting this into a few functions, but I think it can also wait for a future change.

Reply:
I thought about this before. Now I move all non-trivial algorithms to independent functions for readability and add reference comments.